src/main/resources/templates/db_scripts/itemEffectiveLocation.sql

```diff
@@ -0,0 +1,86 @@
+    RETURN NEW;
+  END;
+  $$ LANGUAGE 'plpgsql';
+
+-- Set Item effective location in case insert or update of temporary/permanentLocationId or holdingsRecordId properties
+CREATE OR REPLACE FUNCTION ${myuniversity}_${mymodule}.update_effective_location_on_item_update() RETURNS trigger
+AS $$
+  declare
+    effective_location_id jsonb;
+  begin
+    effective_location_id = coalesce(NEW.jsonb->'temporaryLocationId', NEW.jsonb->'permanentLocationId');
+    if (effective_location_id IS NOT NULL) then
+      NEW.jsonb = jsonb_set(NEW.jsonb, '{effectiveLocationId}', effective_location_id);
+      return NEW;
+    end if;
+    effective_location_id = NEW.jsonb->'effectiveLocationId';
+    if (effective_location_id IS NOT NULL) then
+      -- Field exists so we know it was set by the holdings trigger
+      -- because RMB always removes this read-only field.
+      if (effective_location_id ? '') then
+        -- Holdings trigger uses empty string to signal removal, so there is no need to run a holdings query.
+        -- "effectiveLocationId" must be a valid UUID or removed at all, null or empty string is not allowed.
+        NEW.jsonb = NEW.jsonb - 'effectiveLocationId';
+      end if;
+      return NEW;
+    end if;
+    -- No change in item's sensitive fields?
+    if (TG_OP = 'UPDATE'
+        AND (OLD.jsonb->'holdingsRecordId' IS NOT DISTINCT FROM NEW.jsonb->'holdingsRecordId')
+        AND OLD.jsonb->'temporaryLocationId' IS NULL
+        AND OLD.jsonb->'permanentLocationId' IS NULL) then
+      -- No change in the relevant fields, we can use the old value and avoid a holdings query.
+      effective_location_id = OLD.jsonb->'effectiveLocationId';
+      if (effective_location_id IS NOT NULL) then
+        -- We need to add it again because RMB always removes read-only fields.
+        NEW.jsonb = jsonb_set(NEW.jsonb, '{effectiveLocationId}', effective_location_id);
+      end if;
+      return NEW;
+    end if;
+    SELECT coalesce(jsonb->'temporaryLocationId', jsonb->'permanentLocationId')
+      INTO effective_location_id
+      FROM ${myuniversity}_${mymodule}.holdings_record
+      WHERE id = (NEW.jsonb->>'holdingsRecordId')::uuid
+      LIMIT 1;
+    if (effective_location_id IS NOT NULL) then
+      NEW.jsonb = jsonb_set(NEW.jsonb, '{effectiveLocationId}', effective_location_id);
+    end if;
+    return NEW;
+  END;
+$$ LANGUAGE 'plpgsql';
+
+DROP TRIGGER IF EXISTS update_effective_location_for_items ON ${myuniversity}_${mymodule}.holdings_record;
+-- should be after update trigger, will allow item trigger to fetch up-to-date locations from holding
+create trigger update_effective_location_for_items after update
+on ${myuniversity}_${mymodule}.holdings_record for each row execute procedure ${myuniversity}_${mymodule}.update_effective_location_on_holding_update();
+
+
+DROP TRIGGER IF EXISTS update_effective_location ON ${myuniversity}_${mymodule}.item;
+create trigger update_effective_location before insert or update
```

Comment:
Why is this defined as before insert or update and the holdings trigger defined as after update?

Reply:
We need to calculate the effectiveLocationId for the newly created items + update the property when one (or several) of the sensitive fields have been changed. 

And for holdings we have to recalculate item's effective location on update only. I've declared it as after update in order to allow item's trigger to read new values from holdings table (but I'm not sure if this still required after @julianladisch rework).

Reply:
The holdings trigger should be defined as an "after" trigger because it doesn't change the holdings record. It can wait until all checks and holdings "before" triggers have run and did not cancel the update.
The item trigger should be defined as an "before" trigger because it changes the item record by setting the effective location id. This should be done before the item is saved so that all item "after" triggers have access to this new effective value.

Reply:
Thanks for the explanation @julianladisch